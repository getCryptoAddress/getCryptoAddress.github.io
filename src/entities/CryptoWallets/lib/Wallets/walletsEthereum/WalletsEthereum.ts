import WalletsAbstract from "@/entities/CryptoWallets/lib/Wallets/WalletsAbstract";
import generatePrivateKey from "@/entities/CryptoWallets/lib/PrivateKeys/generatePrivateKey";
import convertUint8ArrayToHex from "@/entities/CryptoWallets/lib/PrivateKeys/convertUint8ArrayToHex";
import type {
  EthereumPrivateKey,
  EthereumWalletPayload,
} from "./WalletsEthereum.types";
import type { Wallet } from "@/entities/CryptoWallets/lib/Wallets/useWallet.types";
import validatePrivateKey from "@/entities/CryptoWallets/lib/PrivateKeys/validatePrivateKey";
import { publicKeyCreate } from "ethereum-cryptography/secp256k1-compat";
import { keccak256 } from "ethereum-cryptography/keccak";

export default class WalletsEthereum extends WalletsAbstract<
  EthereumWalletPayload,
  EthereumPrivateKey
> {
  async makeWallet(
    payload: EthereumWalletPayload,
    initialPrivateKey?: EthereumPrivateKey,
  ): Promise<Wallet> {
    const privateKey = await this.#getPrivateKey(initialPrivateKey);

    const publicKey = publicKeyCreate(privateKey, false).slice(1); // Remove the 04 prefix
    const address = keccak256(publicKey).slice(-20);

    return {
      privateKey: this.#getFormattedPrivateKey(privateKey),
      address: this.#getFormattedAddress(address),
    };
  }

  async #getPrivateKey(initialPrivateKey?: EthereumPrivateKey) {
    if (initialPrivateKey) {
      if (!(await validatePrivateKey(initialPrivateKey, 32))) {
        throw new Error("Invalid private key");
      }
      return initialPrivateKey;
    }
    return generatePrivateKey(32);
  }

  #getFormattedPrivateKey(key: Uint8Array): string {
    return "0x" + convertUint8ArrayToHex(key);
  }
  #getFormattedAddress(key: Uint8Array): string {
    return "0x" + convertUint8ArrayToHex(key);
  }
}
