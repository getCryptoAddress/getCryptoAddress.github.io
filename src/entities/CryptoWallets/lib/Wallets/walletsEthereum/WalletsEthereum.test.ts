import { expect, suite, test } from "vitest";
import WalletsEthereum from "./WalletsEthereum";

const key1 = new Uint8Array([
  132, 205, 117, 44, 180, 113, 245, 200, 213, 19, 97, 187, 70, 157, 86, 111, 19,
  131, 14, 18, 196, 57, 9, 203, 64, 196, 19, 26, 73, 35, 126, 80,
]);
const key2 = new Uint8Array([
  133, 205, 117, 44, 180, 113, 245, 200, 213, 19, 97, 187, 70, 157, 86, 111, 19,
  131, 14, 18, 196, 57, 9, 203, 64, 196, 19, 26, 73, 35, 126, 80,
]);

const notValidKey1 = new Uint8Array([
  255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
  255, 19, 131, 14, 18, 196, 57, 9, 203, 64, 196, 19, 26, 73, 35, 126, 80,
]);

const notValidKey2 = new Uint8Array([
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1,
]);

suite("WalletsEthereum", () => {
  test("should generate wallet correctly", async () => {
    const walletGenerator = new WalletsEthereum();
    const wallet1 = await walletGenerator.makeWallet(null, key1);
    expect(JSON.stringify(wallet1)).toMatchSnapshot();

    const wallet2 = await walletGenerator.makeWallet(null, key2);
    expect(JSON.stringify(wallet2)).toMatchSnapshot();
  });

  test("should generate wallet correctly without initialPrivateKey", async () => {
    const walletGenerator = new WalletsEthereum();

    const wallet = await walletGenerator.makeWallet(null);

    expect(typeof wallet.address).toBe("string");
    expect(wallet.address.length).not.toBe(0);
    expect(typeof wallet.privateKey).toBe("string");
    expect(wallet.privateKey.length).not.toBe(0);
  });

  test("should throw error for invalid initialPrivateKey", async () => {
    const walletGenerator = new WalletsEthereum();

    await expect(
      walletGenerator.makeWallet(null, notValidKey1),
    ).rejects.toThrow("Invalid private key");
    await expect(
      walletGenerator.makeWallet(null, notValidKey2),
    ).rejects.toThrow("offset is out of bounds");
  });
});
