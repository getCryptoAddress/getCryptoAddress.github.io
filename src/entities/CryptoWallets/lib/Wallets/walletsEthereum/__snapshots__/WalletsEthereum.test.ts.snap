// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WalletsEthereum > should generate wallet correctly 1`] = `"{"privateKey":"0x84cd752cb471f5c8d51361bb469d566f13830e12c43909cb40c4131a49237e50","address":"0x4a1bab2023db29cc924e475ceffafb81645f1ab5"}"`;

exports[`WalletsEthereum > should generate wallet correctly 2`] = `"{"privateKey":"0x85cd752cb471f5c8d51361bb469d566f13830e12c43909cb40c4131a49237e50","address":"0xd715914e3592c611d0614e4ad53bed7a09755326"}"`;
